### PR TITLE
Docker api 1.24

### PR DESCRIPTION
**Pull request by shadeslayer**
Move to Docker API 1.24 to support privileged containers

### CODE REVIEW

**Comment by bhush9** [comment]
Looks good to me at first glance, but I don't know the code full enough.. so +1 but wait for +2 from someone else.

File: lib/ci/container.rb [hunk 1]
```diff
@@ -33,11 +35,11 @@ def self.create(connection = Docker.connection,
                     **options_)
       # FIXME: commented to allow tests passing with old containment data
       # assert_version
+      @binds = binds
```
Comment (apachelogger) [inline lib/ci/container.rb]: this is a class instance variable, previously it was an instance variable. this effectively makes it impossible to have two containers at the same time as they will have the same binds

File: lib/ci/container.rb [hunk 2]
```diff
@@ -33,11 +35,11 @@ def self.create(connection = Docker.connection,
...
       options = merge_env(default_create_options, options_)
       options = options.merge(options_)
       options = override_options(options, name, binds)
       c = super(options, connection)
-      c.send(:instance_variable_set, :@binds, binds)
```
Comment (apachelogger) [inline lib/ci/container.rb]: regression as mentioned above

File: lib/ci/container/ephemeral.rb [hunk 3]
```diff
@@ -17,10 +17,13 @@ class << self
 
     def stop(options = {})
       super(options)
-      kill!(options)
+      kill!(options) if running?
       rescued_remove
     end
 
+    def running?
+      json['State']['Running']
```
Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: what if it is paused?

File: lib/ci/container.rb [hunk 4]
```diff
@@ -1,5 +1,7 @@
 require 'docker'
 
+Docker::API_VERSION = '1.24'.freeze
```
Comment (apachelogger) [inline lib/ci/container.rb]: This seems folly. If anything it should set the version only if it is not already >=1.24

File: test/test_ci_containment.rb [hunk 5]
```diff
@@ -64,14 +39,16 @@ def setup
         }
         config.filter_sensitive_data('<%= Dir.pwd %>', :erb_pwd) { Dir.pwd }
       end
-      # Chdir to root, as Containment will set the working dir to PWD and this
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: Why was the comment removed?

File: test/test_ci_containment.rb [hunk 6]
```diff
@@ -276,22 +253,18 @@ def test_image_is_pangeaimage
...
-        vcr_it(__method__, erb: true, tag: :erb_pwd) do
-          c = Containment.new(@job_name, image: @image, binds: [Dir.pwd])
-          c.run(Cmd: ['bash' '-c', 'exit 0'])
-          args = InterceptStartContainer.intercept_start_args
-          assert_equal(1, args.size)
-          args = args[0]
-          assert_equal(CI::Container::DirectBindingArray.to_bindings([Dir.pwd]),
-                       args[:Binds])
+      # sure the binds we put in a re the binds that are passed to docker.
+      VCR.turned_off do
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: Why are network transactions enabled here?

File: test/test_ci_containment.rb [hunk 7]
```diff
@@ -276,22 +253,18 @@ def test_image_is_pangeaimage
...
+        Dir.chdir(@tmpdir) do
+          assert_raise CI::BindsPassed do
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: raise should come out of run and nothing else?!

File: test/test_ci_containment.rb [hunk 8]
```diff
@@ -64,14 +39,16 @@ def setup
...
-      # is slightly unwanted for tmpdir tests.
       Dir.chdir('/')
 
       @job_name = 'vivid_unstable_test'
       @image = PangeaImage.new('ubuntu', 'vivid')
 
-      VCR.turned_off { cleanup_container }
+      VCR.turned_off do
+        cleanup_container
+        image = Docker::Image.create(fromImage: 'ubuntu:vivid')
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: Image creation and tagging should not be done when in playback mode. (see vcr_it in container upgrade test or smth)

File: test/test_ci_container.rb [hunk 9]
```diff
@@ -29,13 +29,12 @@ def setup
...
-    # Chdir to root, as Containment will set the working dir to PWD and this
-    # is slightly unwanted for tmpdir tests.
-    Dir.chdir('/')
-
     @job_name = self.class.to_s
     @image = 'ubuntu:15.04'
-    VCR.turned_off { cleanup_container }
+    VCR.turned_off do
+      cleanup_container
+      Docker::Image.create(fromImage: @image)
```
Comment (apachelogger) [inline test/test_ci_container.rb]: Image creation should not be done when in playback mode. (see vcr_it in container upgrade test or smth)

File: lib/ci/container/ephemeral.rb (hunk 3 above)
Comment (shadeslayer) [inline lib/ci/container/ephemeral.rb]: Docker reports it as Paused but running : 

```json
{"Status"=>"paused", "Running"=>true, "Paused"=>true, "Restarting"=>false, "OOMKilled"=>false, "Dead"=>false, "Pid"=>28607, "ExitCode"=>0, "Error"=>"", "StartedAt"=>"2016-10-13T14:52:08.178882021Z", "FinishedAt"=>"0001-01-01T00:00:00Z"} 
```

Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: well, you probably should raise if it paused to make it obvious what the problem is

Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: Oh, and on a related note. Change this to use fetch(), [] is bullshit.

File: ci-tooling/data/projects [hunk 10]
```diff
@@ -1 +1 @@
-Subproject commit f695890e4bca5935f9e41252a9bc0e84739788f1
+Subproject commit 1c2ffd1802da4ce9280d07a13e051dbf14410e3b
```
Comment (apachelogger) [inline ci-tooling/data/projects]: hachoo

File: lib/ci/container.rb [hunk 11]
```diff
@@ -1,5 +1,7 @@
 require 'docker'
 
+Docker::API_VERSION = '1.24'.freeze if Docker::API_VERSION.to_f < 1.24
```
Comment (apachelogger) [inline lib/ci/container.rb]: http://stackoverflow.com/questions/2051229/how-to-compare-versions-in-ruby

File: test/test_ci_containment.rb [hunk 12]
```diff
@@ -276,22 +257,19 @@ def test_image_is_pangeaimage
...
     def test_ZZZ_binds # Last test always! Changes VCR configuration.
       # Container binds were overwritten by Containment at some point, make
-      # sure the binds we put in are the binds that are passed to docker.
-      Dir.chdir(@tmpdir) do
-        Docker::Container.prepend(InterceptStartContainer)
-        InterceptStartContainer.intercept_start_container = true
-        vcr_it(__method__, erb: true, tag: :erb_pwd) do
+      # sure the binds we put in a re the binds that are passed to docker.
+      WebMock.disable_net_connect!
+      VCR.turned_off do
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: This still makes no sense.
Just don't turn off vcr?

File: test/test_ci_container.rb [hunk 13]
```diff
@@ -14,7 +14,7 @@ def cleanup_container
     # the vcr data.
     c = Docker::Container.get(@job_name)
     c.stop
-    c.kill!
+    c.kill! if c.json['State']['Running']
```
Comment (apachelogger) [inline test/test_ci_container.rb]: fetch

File: test/test_ci_containment.rb [hunk 14]
```diff
@@ -43,7 +18,7 @@ def cleanup_container
       # the vcr data.
       c = Docker::Container.get(@job_name)
       c.stop
-      c.kill!
+      c.kill! if c.json['State']['Running']
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: fetch

File: ci-tooling/data/projects (hunk 10 above)
Comment (shadeslayer) [inline ci-tooling/data/projects]: I have no clue why that's in there, will remove when rebasing

File: test/test_ci_containment.rb (hunk 12 above)
Comment (shadeslayer) [inline test/test_ci_containment.rb]: It needs turning off for some reason, because without it :

```
Error: test_ZZZ_binds(CI::ContainmentTest):
  VCR::Errors::UnhandledHTTPRequestError: 
  
  ================================================================================
  An HTTP request has been made that VCR does not know how to handle:
    GET unix:///var/run/docker.sock/v1.24/version
    Body: 
```

Comment (apachelogger) [inline test/test_ci_containment.rb]: mock the version query?

if you turn vcr off and disable net connect that would fail on that /version access!

File: lib/ci/container.rb [hunk 15]
```diff
@@ -1,5 +1,9 @@
 require 'docker'
 
+if Gem::Version.new(Docker::API_VERSION) < Gem::Version.new(1.24)
```
Comment (apachelogger) [inline lib/ci/container.rb]: Actually. put that in a docker.rb along with the `require 'docker'` then require_relative

File: lib/ci/container.rb [hunk 16]
```diff
@@ -62,29 +65,22 @@ def start(options = {})
...
       # - Env: Sensible defaults for LANG, PATH, DEBIAN_FRONTEND
+      # - Binds: Set to binds used on {create}
+      # - Ulimits: Set to sane defaults with lower nofile property
       # @return [Hash]
       def default_create_options
         {
+          # Force standard ulimit in the container.
+          # Otherwise pretty much all APT IO operations are insanely slow:
+          # https://bugs.launchpad.net/ubuntu/+source/apt/+bug/1332440
+          # This in particular affects apt-extracttemplates which will take up to
```
Comment (apachelogger) [inline lib/ci/container.rb]: rubocop

File: lib/ci/container/ephemeral.rb [hunk 17]
```diff
@@ -4,6 +4,7 @@ module CI
   # An ephemeral container. It gets automatically removed after it closes.
   # This is slightly more reliable than Docker's own implementation as
   # this goes to extra lengths to make sure the container disappears.
+  class EphemeralContainerUnhandledState < StandardError; end
   class EphemeralContainer < Container
```
Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: rubocop

File: lib/ci/container/ephemeral.rb [hunk 18]
```diff
@@ -17,10 +18,15 @@ class << self
...
     def stop(options = {})
       super(options)
-      kill!(options)
+      kill!(options) if running?
       rescued_remove
     end
 
+    def running?
+      state = json.fetch('State')
+      raise EphemeralContainerUnhandledState unless %w(exited running).include? state.fetch('Status')
```
Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: rubocop

also, this is code and not a dsl so use brackets

also, comment on why this is necessary

also, so, I am not allowed to have containers on my local system if I want to be able to run the pangea tests?

File: lib/ci/docker.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+require 'docker'
+
+if Gem::Version.new(Docker::API_VERSION) < Gem::Version.new(1.24)
+  Docker::API_VERSION = '1.24'.freeze
```
Comment (apachelogger) [inline lib/ci/docker.rb]: I just noticed. this coerces the version regardless of whether the server actually supports that version. it takes no measures if that is not the case

File: lib/ci/directbindingarray.rb [hunk 20]
```diff
@@ -21,6 +21,7 @@ def self.to_volumes(array)
 
     # @return [Array] Binds API array of the form ["Path:Path"]
     def self.to_bindings(array)
+      array ||= []
```
Comment (apachelogger) [inline lib/ci/directbindingarray.rb]: one would argue that this method should not be called with anything but an array and if one calls it with anything but an array that is a bug in the calling code

also if anything the code should read `return [] unless array` as this very line is in a very explicit fashion rendering the collect {} noop, as such initializing it to empty and then running collect on an empty array collecting another empty array is bullshit of the highest order

File: test/test_ci_container.rb [hunk 21]
```diff
@@ -27,23 +28,36 @@ def setup
...
+    VCR.turned_off do
+      cleanup_container
+    end
   end
 
   def teardown
     VCR.turned_off { cleanup_container }
   end
 
+  # FIXME: Make vcr_it a common method for tests
```
Comment (apachelogger) [inline test/test_ci_container.rb]: resolve the fixme?

File: test/test_ci_containment.rb [hunk 22]
```diff
@@ -102,7 +76,14 @@ def assert_handler_not_set(signal)
 
     def vcr_it(meth, **kwords)
       VCR.use_cassette(meth, kwords) do |cassette|
-        CI::EphemeralContainer.safety_sleep = 0 unless cassette.recording?
+        if cassette.recording?
+          VCR.turned_off do
+            image = Docker::Image.create(fromImage: 'ubuntu:vivid')
+            image.tag(repo: @image.repo, tag: @image.tag) unless Docker::Image.exist? @image.to_s
```
Comment (apachelogger) [inline test/test_ci_containment.rb]: method calls use brackets.

File: lib/ci/container.rb [hunk 23]
```diff
@@ -62,29 +60,22 @@ def start(options = {})
...
       # @return [Hash]
       def default_create_options
         {
+          # Force standard ulimit in the container.
+          # Otherwise pretty much all APT IO operations are insanely slow:
+          # https://bugs.launchpad.net/ubuntu/+source/apt/+bug/1332440
+          # This in particular affects apt-extracttemplates which will take up
+          # to 20 minutes where it should take maybe 1/10 of that.
+          Ulimits: [{ Name: 'nofile', Soft: 1024, Hard: 1024 }],
+          Binds: DirectBindingArray.to_bindings(@binds),
```
Comment (apachelogger) [inline lib/ci/container.rb]: /tmp/pangea-tooling/lib/ci/container.rb:78: warning: instance variable @binds not initialized

Comment (shadeslayer) [inline lib/ci/container.rb]: Ah the whole binds stuff is a bit silly right now, I'll look into fixing this.

File: test/test_ci_container.rb (hunk 21 above)
Comment (shadeslayer) [inline test/test_ci_container.rb]: Will resolve at a later date.

Comment (apachelogger) [inline test/test_ci_container.rb]: Or just do it now while you still have the code in mind.

Comment (shadeslayer) [inline test/test_ci_container.rb]: I only just realized, can't be made common, one might want to do some sort of test setup for recording cassettes, so I guess I should drop the fixme

File: test/test_ci_container.rb [hunk 24]
```diff
@@ -27,23 +28,36 @@ def setup
...
+      config.filter_sensitive_data('<%= Dir.pwd %>', :erb_pwd) { Dir.pwd }
     end
 
-    # Chdir to root, as Containment will set the working dir to PWD and this
-    # is slightly unwanted for tmpdir tests.
-    Dir.chdir('/')
-
     @job_name = self.class.to_s
     @image = 'ubuntu:15.04'
-    VCR.turned_off { cleanup_container }
```
Comment (apachelogger) [inline test/test_ci_container.rb]: why was this changed?

Comment (shadeslayer) [inline test/test_ci_container.rb]: Probably because I was doing more stuff with VCR turned off, and then we changed things to not do multi line things anymore. So I've reverted the change

File: lib/ci/container/ephemeral.rb [hunk 25]
```diff
@@ -17,10 +19,16 @@ class << self
...
     def stop(options = {})
       super(options)
-      kill!(options)
+      kill!(options) if running?
       rescued_remove
     end
 
+    def running?
+      state = json.fetch('State')
+      raise EphemeralContainerUnhandledState unless
```
Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: unless LONGEXP
  raise Error
end

File: lib/ci/container/ephemeral.rb [hunk 26]
```diff
@@ -1,11 +1,13 @@
 require_relative '../container'
 
 module CI
+  class EphemeralContainerUnhandledState < StandardError; end
```
Comment (apachelogger) [inline lib/ci/container/ephemeral.rb]: I'd move that into the class

`CI::EphemeralContainer::UnhandledState` is more readable than `CI::EphemeralContainerUnhandledState`

File: test/test_ci_container.rb [hunk 27]
```diff
@@ -42,8 +40,22 @@ def teardown
     VCR.turned_off { cleanup_container }
   end
 
+  # FIXME: Make vcr_it a common method for tests
```
Comment (apachelogger) [inline test/test_ci_container.rb]: fixme

**Comment by bhush9** [comment]
I suppose this can be closed now that is merged
